### ISSUE TITLE
Re07-usepopcorn > src > App, useMovies | custom hook set up error

### ISSUE THREAD

**Issue by devYuraKim** [devYuraKim/react#10]
I wasn't so sure how to handle `query` and `setSelectedId`, so I just passed them in as arguments from `App` component to `useMovies` hook, then got this error message: 

> _hooks_useMovies__WEBPACK_IMPORTED_MODULE_1__ is not a function. (In '_hooks_useMovies__WEBPACK_IMPORTED_MODULE_1__(query, setSelectedId)', '_hooks_useMovies__WEBPACK_IMPORTED_MODULE_1__' is "/static/media/useMovies.df84d1619faf16246f81")

**App component**
```  
const { movies, isLoading, error } = useMovies(query, setSelectedId); 
```

**useMovies hook**
```
const KEY = "d9c29e1c";

export default function useMovies(query, setSelectedId) {
  const [movies, setMovies] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState("");

  useEffect(
    function () {
      const controller = new AbortController();
      async function fetchMovies() {
        try {
          setError(null);
          setIsLoading(true);
          const res = await fetch(
            `https://www.omdbapi.com/?apikey=${KEY}&s=${query}`,
            { signal: controller.signal }
          );
          if (!res.ok) throw new Error("Lost Connection");

          const data = await res.json();
          if (data.Response === "False") throw new Error("No Result");

          setMovies(data.Search);
          setIsLoading(false);
          setError(null);
        } catch (err) {
          if (err.name !== "AbortError") {
            console.log(err.message);
            setError(err.message);
          }
        } finally {
          setIsLoading(false);
        }
      }
      if (!query.length) {
        setMovies([]);
        setError(null);
        return;
      }
      setSelectedId(null);
      fetchMovies();

      return function () {
        controller.abort();
      };
    },
    [query, setSelectedId]
  );

  return [movies, isLoading, error];
}

```

**Comment by devYuraKim** [comment]
TL;DR: not arguments, but extension
---
file extension .js was missing for useMovies.
passing `query` and `setSelectedId` worked fine.

**Comment by devYuraKim** [comment]
New Issue: No result for the fetch request
---

> TypeError: undefined is not an object (evaluating 'movies.length')

**Comment by devYuraKim** [comment]
custom hooks return `{object}` not `[array]`
---
**STUPID ME! 
The problem all along was that I was returning an `[array]` instead of an `{object}`.**

Initially, I thought the issue was with the `<NumResults>` component where `movies.length` was. To troubleshoot, I commented out the entire component and ran the application. At first glance, it seemed to work fine.

However, when I typed in a search keyword, nothing happened. I only saw the `"No Result"` message in the console, which is triggered when `data.Response` equals `"False"`. I initially suspected that setSelectedId might have been interfering with the data fetching.

To investigate further, I checked the `Network` tab and confirmed that data was being fetched. Since the `query` was initially set to `null`, it made sense to get `"No Result"` when the app mounted. As soon as I entered a search keyword, the query was updated, and the corresponding data was successfully fetched.

I was really stuck, so I decided to compare my code with the working version line by line. Even then, I couldn’t find the issue. I thought about taking a break and coming back to it later. 

As a last resort, I double-checked Q&A to see if anyone else had encountered the same problem. Unfortunately, no one had. However, I came across a question about which brackets to use when returning values. It then struck me that I might have used the wrong brackets— and I did!

**Comment by devYuraKim** [comment]
TL;DR: custom hook can return both `{object}` and `[array]`
---
OK, I must admit that I'm quite embarrassed about not getting this sooner, but I’m taking comfort in the fact that this is all part of the learning process...

- `Array ([])`: Suitable for simple hooks with a fixed number of return values, especially when mimicking the pattern of `useState` or `useReducer`. The order of values is understood through context and is straightforward to use.

- `Object ({})`: Provides `named properties`, making it easier to understand what each value represents, especially as the number of return values increases or if the values are more complex.
